# Playstore Data Analysis

##  Importing libraries

In [228]:
import pandas as pd #data handling
import matplotlib.pyplot as plt #visualization
import seaborn as sns #visualization

## loading Data

In [229]:
df=pd.read_csv("Playstore.csv")
df

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10836,Sya9a Maroc - FR,FAMILY,4.5,38,53M,"5,000+",Free,0,Everyone,Education,"July 25, 2017",1.48,4.1 and up
10837,Fr. Mike Schmitz Audio Teachings,FAMILY,5.0,4,3.6M,100+,Free,0,Everyone,Education,"July 6, 2018",1,4.1 and up
10838,Parkinson Exercices FR,MEDICAL,NaN,3,9.5M,"1,000+",Free,0,Everyone,Medical,"January 20, 2017",1,2.2 and up
10839,The SCP Foundation DB fr nn5n,BOOKS_AND_REFERENCE,4.5,114,Varies with device,"1,000+",Free,0,Mature 17+,Books & Reference,"January 19, 2015",Varies with device,Varies with device


## Inspect Data

In [230]:
df.head()


,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up


In [231]:
df.tail()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
10836,Sya9a Maroc - FR,FAMILY,4.5,38,53M,"5,000+",Free,0,Everyone,Education,"July 25, 2017",1.48,4.1 and up
10837,Fr. Mike Schmitz Audio Teachings,FAMILY,5.0,4,3.6M,100+,Free,0,Everyone,Education,"July 6, 2018",1,4.1 and up
10838,Parkinson Exercices FR,MEDICAL,NaN,3,9.5M,"1,000+",Free,0,Everyone,Medical,"January 20, 2017",1,2.2 and up
10839,The SCP Foundation DB fr nn5n,BOOKS_AND_REFERENCE,4.5,114,Varies with device,"1,000+",Free,0,Mature 17+,Books & Reference,"January 19, 2015",Varies with device,Varies with device
10840,iHoroscope - 2018 Daily Horoscope & Astrology,LIFESTYLE,4.5,398307,19M,"10,000,000+",Free,0,Everyone,Lifestyle,"July 25, 2018",Varies with device,Varies with device


In [232]:
df.sample()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
3454,Google Drive,PRODUCTIVITY,4.4,2731171,Varies with device,"1,000,000,000+",Free,0,Everyone,Productivity,"August 6, 2018",Varies with device,Varies with device


## Selecting Rows

In [233]:
df.iloc[9999]

App                 EW Handbook
Category               BUSINESS
Rating                      NaN
Reviews                       1
Size                        17M
Installs                   100+
Type                       Free
Price                         0
Content Rating         Everyone
Genres                 Business
Last Updated      June 11, 2015
Current Ver                   4
Android Ver        2.3.3 and up
Name: 9999, dtype: object

In [234]:
df.iloc[5:11]

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
5,Paper flowers instructions,ART_AND_DESIGN,4.4,167,5.6M,"50,000+",Free,0,Everyone,Art & Design,"March 26, 2017",1,2.3 and up
6,Smoke Effect Photo Maker - Smoke Editor,ART_AND_DESIGN,3.8,178,19M,"50,000+",Free,0,Everyone,Art & Design,"April 26, 2018",1.1,4.0.3 and up
7,Infinite Painter,ART_AND_DESIGN,4.1,36815,29M,"1,000,000+",Free,0,Everyone,Art & Design,"June 14, 2018",6.1.61.1,4.2 and up
8,Garden Coloring Book,ART_AND_DESIGN,4.4,13791,33M,"1,000,000+",Free,0,Everyone,Art & Design,"September 20, 2017",2.9.2,3.0 and up
9,Kids Paint Free - Drawing Fun,ART_AND_DESIGN,4.7,121,3.1M,"10,000+",Free,0,Everyone,Art & Design;Creativity,"July 3, 2018",2.8,4.0.3 and up
10,Text on Photo - Fonteee,ART_AND_DESIGN,4.4,13880,28M,"1,000,000+",Free,0,Everyone,Art & Design,"October 27, 2017",1.0.4,4.1 and up


## Checking Data

In [235]:
df.dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

In [236]:
df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [237]:
df.shape

(10841, 13)

In [238]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


- Rating column 1574 missing count
- Type column 1 missing count
- Content Rating column 1 missing count
- Current Ver column 8 missing count
- Android Ver column 3 missing count

# Single column

In [239]:
df['App']

0           Photo Editor & Candy Camera & Grid & ScrapBook
1                                      Coloring book moana
2        U Launcher Lite – FREE Live Cool Themes, Hide ...
3                                    Sketch - Draw & Paint
4                    Pixel Draw - Number Art Coloring Book
                               ...                        
10836                                     Sya9a Maroc - FR
10837                     Fr. Mike Schmitz Audio Teachings
10838                               Parkinson Exercices FR
10839                        The SCP Foundation DB fr nn5n
10840        iHoroscope - 2018 Daily Horoscope & Astrology
Name: App, Length: 10841, dtype: object

In [240]:
df['Category']

0             ART_AND_DESIGN
1             ART_AND_DESIGN
2             ART_AND_DESIGN
3             ART_AND_DESIGN
4             ART_AND_DESIGN
                ...         
10836                 FAMILY
10837                 FAMILY
10838                MEDICAL
10839    BOOKS_AND_REFERENCE
10840              LIFESTYLE
Name: Category, Length: 10841, dtype: object

In [241]:
df['Price']

0        0
1        0
2        0
3        0
4        0
        ..
10836    0
10837    0
10838    0
10839    0
10840    0
Name: Price, Length: 10841, dtype: object

# Multiple column

In [242]:
df[['App','Category']]

,App,Category
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN
1,Coloring book moana,ART_AND_DESIGN
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN
3,Sketch - Draw & Paint,ART_AND_DESIGN
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN
...,...,...
10836,Sya9a Maroc - FR,FAMILY
10837,Fr. Mike Schmitz Audio Teachings,FAMILY
10838,Parkinson Exercices FR,MEDICAL
10839,The SCP Foundation DB fr nn5n,BOOKS_AND_REFERENCE


In [243]:
New = df[['App','Category','Price']]

## Filter the Data

In [244]:
df[df['Rating']<3]

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
477,Calculator,DATING,2.6,57,6.2M,"1,000+",Paid,$6.99,Everyone,Dating,"October 25, 2017",1.1.6,4.0 and up
518,Just She - Top Lesbian Dating,DATING,1.9,953,19M,"100,000+",Free,0,Mature 17+,Dating,"July 18, 2018",6.3.7,5.0 and up
520,EliteSingles – Dating for Single Professionals,DATING,2.5,5377,19M,"500,000+",Free,0,Mature 17+,Dating,"July 31, 2018",4.8.5,4.0.3 and up
527,Sugar Daddy Dating App,DATING,2.5,277,5.7M,"100,000+",Free,0,Mature 17+,Dating,"December 4, 2017",3.0.0,4.1 and up
528,Adult Dirty Emojis,DATING,2.8,80,5.5M,"10,000+",Free,0,Teen,Dating,"November 6, 2017",1,4.0.3 and up
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10624,Employ Florida Mobile,BUSINESS,2.9,97,9.9M,"10,000+",Free,0,Everyone,Business,"June 13, 2018",4.5.9,4.0.3 and up
10665,SB · FN 1870 Mobile Banking,FINANCE,2.9,139,3.3M,"10,000+",Free,0,Everyone,Finance,"June 19, 2017",3.0.5,4.0 and up
10677,Pint - FN Theme,PERSONALIZATION,2.5,6,234k,100+,Free,0,Everyone,Personalization,"August 10, 2013",1,2.2 and up
10757,Fisher-Price® Smart Connect™,TOOLS,2.7,422,72M,"50,000+",Free,0,Everyone,Tools,"February 23, 2018",2.4.1,4.4 and up


In [245]:
df[df['Rating']==4]

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
20,Logo Maker - Small Business,ART_AND_DESIGN,4.0,450,14M,"100,000+",Free,0,Everyone,Art & Design,"April 20, 2018",4,4.1 and up
39,How To Color Disney Princess - Coloring Pages,ART_AND_DESIGN,4.0,591,9.4M,"500,000+",Free,0,Everyone,Art & Design,"March 31, 2018",1,4.0 and up
50,Real Tractor Farming,AUTO_AND_VEHICLES,4.0,1598,56M,"1,000,000+",Free,0,Everyone,Auto & Vehicles,"July 26, 2018",11,4.1 and up
58,Restart Navigator,AUTO_AND_VEHICLES,4.0,1403,201k,"100,000+",Free,0,Everyone,Auto & Vehicles,"August 26, 2014",1.0.1,2.2 and up
66,Used Cars Mexico,AUTO_AND_VEHICLES,4.0,190,2.5M,"50,000+",Free,0,Everyone,Auto & Vehicles,"October 14, 2016",1,2.3 and up
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10802,FR: My Famous Lover,FAMILY,4.0,185,28M,"10,000+",Free,0,Teen,Entertainment,"August 6, 2015",1.3.0,3.0 and up
10814,FR: My Secret Pets!,FAMILY,4.0,785,31M,"50,000+",Free,0,Teen,Entertainment,"June 3, 2015",1.3.1,3.0 and up
10817,HTC Sense Input - FR,TOOLS,4.0,885,8.0M,"100,000+",Free,0,Everyone,Tools,"October 30, 2015",1.0.612928,5.0 and up
10826,Frim: get new friends on local chat rooms,SOCIAL,4.0,88486,Varies with device,"5,000,000+",Free,0,Mature 17+,Social,"March 23, 2018",Varies with device,Varies with device


In [246]:
df[df['Content Rating']=='Teen']

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
12,Tattoo Name On My Photo Editor,ART_AND_DESIGN,4.2,44829,20M,"10,000,000+",Free,0,Teen,Art & Design,"April 2, 2018",3.8,4.1 and up
34,I Creative Idea,ART_AND_DESIGN,4.7,353,4.2M,"10,000+",Free,0,Teen,Art & Design,"April 27, 2018",1.6,4.1 and up
72,"Android Auto - Maps, Media, Messaging & Voice",AUTO_AND_VEHICLES,4.2,271920,16M,"10,000,000+",Free,0,Teen,Auto & Vehicles,"July 11, 2018",Varies with device,5.0 and up
112,Selfie Camera Photo Editor & Filter & Sticker,BEAUTY,4.1,187,30M,"50,000+",Free,0,Teen,Beauty,"July 24, 2018",3.0.1,4.0.3 and up
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10804,Poker Pro.Fr,GAME,4.2,5442,17M,"100,000+",Free,0,Teen,Card,"May 22, 2018",4.1.3,2.3 and up
10814,FR: My Secret Pets!,FAMILY,4.0,785,31M,"50,000+",Free,0,Teen,Entertainment,"June 3, 2015",1.3.1,3.0 and up
10819,Fanfic-FR,BOOKS_AND_REFERENCE,3.3,52,3.6M,"5,000+",Free,0,Teen,Books & Reference,"August 5, 2017",0.3.4,4.1 and up
10820,Fr. Daoud Lamei,FAMILY,5.0,22,8.6M,"1,000+",Free,0,Teen,Education,"June 27, 2018",3.8.0,4.1 and up


In [247]:
df[df['Type']=='Paid']

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
234,TurboScan: scan documents and receipts in PDF,BUSINESS,4.7,11442,6.8M,"100,000+",Paid,$4.99,Everyone,Business,"March 25, 2018",1.5.2,4.0 and up
235,Tiny Scanner Pro: PDF Doc Scan,BUSINESS,4.8,10295,39M,"100,000+",Paid,$4.99,Everyone,Business,"April 11, 2017",3.4.6,3.0 and up
290,TurboScan: scan documents and receipts in PDF,BUSINESS,4.7,11442,6.8M,"100,000+",Paid,$4.99,Everyone,Business,"March 25, 2018",1.5.2,4.0 and up
291,Tiny Scanner Pro: PDF Doc Scan,BUSINESS,4.8,10295,39M,"100,000+",Paid,$4.99,Everyone,Business,"April 11, 2017",3.4.6,3.0 and up
427,Puffin Browser Pro,COMMUNICATION,4.0,18247,Varies with device,"100,000+",Paid,$3.99,Everyone,Communication,"July 5, 2018",7.5.3.20547,4.1 and up
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10735,FP VoiceBot,FAMILY,NaN,17,157k,100+,Paid,$0.99,Mature 17+,Entertainment,"November 25, 2015",1.2,2.1 and up
10760,Fast Tract Diet,HEALTH_AND_FITNESS,4.4,35,2.4M,"1,000+",Paid,$7.99,Everyone,Health & Fitness,"August 8, 2018",1.9.3,4.2 and up
10782,Trine 2: Complete Story,GAME,3.8,252,11M,"10,000+",Paid,$16.99,Teen,Action,"February 27, 2015",2.22,5.0 and up
10785,"sugar, sugar",FAMILY,4.2,1405,9.5M,"10,000+",Paid,$1.20,Everyone,Puzzle,"June 5, 2018",2.7,2.3 and up


## Filtering with Multiple Condition

In [248]:
df[(df['Content Rating']=='Teen') & (df['Rating']<4)]

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
152,Google Play Books,BOOKS_AND_REFERENCE,3.9,1433233,Varies with device,"1,000,000,000+",Free,0,Teen,Books & Reference,"August 3, 2018",Varies with device,Varies with device
310,Comic Es - Shojo manga / love comics free of c...,COMICS,3.9,2181,10M,"100,000+",Free,0,Teen,Comics,"March 5, 2018",1.2.12,4.0.3 and up
311,comico Popular Original Cartoon Updated Everyd...,COMICS,3.2,93965,15M,"5,000,000+",Free,0,Teen,Comics,"July 3, 2018",6.3.0,4.0.3 and up
317,Marvel Unlimited,COMICS,3.7,22551,11M,"1,000,000+",Free,0,Teen,Comics,"July 24, 2018",3.16.0,4.4 and up
325,Manga AZ - Manga Comic Reader,COMICS,3.3,125,4.7M,"5,000+",Free,0,Teen,Comics,"July 24, 2018",1.0.4,4.0.3 and up
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10515,Dino Defends king 3 – Dinosaur T rex Hunter Games,GAME,3.6,182,41M,"50,000+",Free,0,Teen,Adventure,"March 1, 2018",2.1,4.1 and up
10518,Block Gun 3D: Haunted Hollow,GAME,3.8,16282,35M,"1,000,000+",Free,0,Teen,Action,"October 9, 2017",1.1.4,2.3 and up
10578,Check Lottery Tickets - Florida,FAMILY,2.0,27,31M,"5,000+",Free,0,Teen,Entertainment,"October 17, 2017",2.1,4.0.3 and up
10782,Trine 2: Complete Story,GAME,3.8,252,11M,"10,000+",Paid,$16.99,Teen,Action,"February 27, 2015",2.22,5.0 and up


In [249]:
df[(df['Content Rating']=='Teen') & (df['Rating']<4) & (df['Category']=='GAME')]

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
1682,Race the Traffic Moto,GAME,3.8,270687,38M,"10,000,000+",Free,0,Teen,Racing,"July 20, 2018",1.0.16,4.0.3 and up
3977,B-Grade Renegade,GAME,3.9,24,64M,"1,000+",Paid,$1.99,Teen,Action,"October 17, 2017",0.8.17,3.2 and up
4847,Route Z,GAME,3.9,24697,16M,"500,000+",Free,0,Teen,Action,"July 10, 2015",1.5,4.0 and up
4901,Assassin's Creed Identity,GAME,3.7,42529,92M,"500,000+",Paid,$1.99,Teen,Action,"November 23, 2016",2.8.2,4.1 and up
5158,AH-1 Viper Cobra Ops - helicopter flight simul...,GAME,3.8,1088,86M,"100,000+",Free,0,Teen,Action,"February 16, 2018",1.0.1,4.3 and up
5241,Blackjack aj Poker,GAME,3.8,8,4.2M,"1,000+",Free,0,Teen,Card,"October 19, 2016",1.1,2.3.3 and up
5293,Counter Terrorist Attack,GAME,3.9,145931,43M,"10,000,000+",Free,0,Teen,Action,"May 31, 2018",5.3.2,2.3 and up
5529,Ghost Snap AR Horror Survival,GAME,3.8,898,33M,"100,000+",Free,0,Teen,Adventure,"December 21, 2016",1.2.3.1,2.3 and up
5636,One Night at Golden Freddy's,GAME,3.3,2717,4.0M,"100,000+",Free,0,Teen,Action,"December 7, 2017",0.1,2.3 and up
5642,Two Nights at jumpscare,GAME,3.8,596,91M,"100,000+",Free,0,Teen,Adventure,"August 17, 2017",1,4.1 and up


## Loc Method

In [250]:
df.loc[7896,'Category']

'FINANCE'

## Rename Column

In [251]:
df.rename(columns={"Content Rating":"Age Category","Type":"Cash Type"},inplace=True)
df

,App,Category,Rating,Reviews,Size,Installs,Cash Type,Price,Age Category,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10836,Sya9a Maroc - FR,FAMILY,4.5,38,53M,"5,000+",Free,0,Everyone,Education,"July 25, 2017",1.48,4.1 and up
10837,Fr. Mike Schmitz Audio Teachings,FAMILY,5.0,4,3.6M,100+,Free,0,Everyone,Education,"July 6, 2018",1,4.1 and up
10838,Parkinson Exercices FR,MEDICAL,NaN,3,9.5M,"1,000+",Free,0,Everyone,Medical,"January 20, 2017",1,2.2 and up
10839,The SCP Foundation DB fr nn5n,BOOKS_AND_REFERENCE,4.5,114,Varies with device,"1,000+",Free,0,Mature 17+,Books & Reference,"January 19, 2015",Varies with device,Varies with device


In [252]:
df['Category'].value_counts()

Category
FAMILY                 1972
GAME                   1144
TOOLS                   843
MEDICAL                 463
BUSINESS                460
PRODUCTIVITY            424
PERSONALIZATION         392
COMMUNICATION           387
SPORTS                  384
LIFESTYLE               382
FINANCE                 366
HEALTH_AND_FITNESS      341
PHOTOGRAPHY             335
SOCIAL                  295
NEWS_AND_MAGAZINES      283
SHOPPING                260
TRAVEL_AND_LOCAL        258
DATING                  234
BOOKS_AND_REFERENCE     231
VIDEO_PLAYERS           175
EDUCATION               156
ENTERTAINMENT           149
MAPS_AND_NAVIGATION     137
FOOD_AND_DRINK          127
HOUSE_AND_HOME           88
AUTO_AND_VEHICLES        85
LIBRARIES_AND_DEMO       85
WEATHER                  82
ART_AND_DESIGN           65
EVENTS                   64
PARENTING                60
COMICS                   60
BEAUTY                   53
1.9                       1
Name: count, dtype: int64

In [253]:
df['Age Category'].value_counts()

Age Category
Everyone           8714
Teen               1208
Mature 17+          499
Everyone 10+        414
Adults only 18+       3
Unrated               2
Name: count, dtype: int64

In [254]:
df['Rating'].value_counts(5)

Rating
4.4     0.118394
4.3     0.114871
4.5     0.110815
4.2     0.101633
4.6     0.087862
4.1     0.075584
4.0     0.060638
4.7     0.053272
3.9     0.041208
3.8     0.032348
5.0     0.029252
3.7     0.025515
4.8     0.024981
3.6     0.018576
3.5     0.017402
3.4     0.013665
3.3     0.010889
4.9     0.009288
3.0     0.008861
3.1     0.007366
3.2     0.006832
2.9     0.004804
2.8     0.004484
2.6     0.002669
2.7     0.002669
2.5     0.002242
2.3     0.002135
2.4     0.002028
1.0     0.001708
2.2     0.001495
1.9     0.001388
2.0     0.001281
1.7     0.000854
1.8     0.000854
2.1     0.000854
1.6     0.000427
1.4     0.000320
1.5     0.000320
1.2     0.000107
19.0    0.000107
Name: proportion, dtype: float64

In [255]:
df['Installs']

0            10,000+
1           500,000+
2         5,000,000+
3        50,000,000+
4           100,000+
            ...     
10836         5,000+
10837           100+
10838         1,000+
10839         1,000+
10840    10,000,000+
Name: Installs, Length: 10841, dtype: object

## Clean Numerical columns with Characters

In [256]:
df['Installs'] = pd.to_numeric(
    df['Installs']
    .astype(str)
    .str.replace(r'[^0-9.-]','',regex=True)
    .replace('',None),
    errors='coerce'
)
df

,App,Category,Rating,Reviews,Size,Installs,Cash Type,Price,Age Category,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,10000.0,Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,500000.0,Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,5000000.0,Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,50000000.0,Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,100000.0,Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10836,Sya9a Maroc - FR,FAMILY,4.5,38,53M,5000.0,Free,0,Everyone,Education,"July 25, 2017",1.48,4.1 and up
10837,Fr. Mike Schmitz Audio Teachings,FAMILY,5.0,4,3.6M,100.0,Free,0,Everyone,Education,"July 6, 2018",1,4.1 and up
10838,Parkinson Exercices FR,MEDICAL,NaN,3,9.5M,1000.0,Free,0,Everyone,Medical,"January 20, 2017",1,2.2 and up
10839,The SCP Foundation DB fr nn5n,BOOKS_AND_REFERENCE,4.5,114,Varies with device,1000.0,Free,0,Mature 17+,Books & Reference,"January 19, 2015",Varies with device,Varies with device


In [257]:
df['Reviews'] = pd.to_numeric(
    df['Reviews']
    .astype(str)
    .str.replace(r'[^0-9.-]','',regex=True)
    .replace('',None),
    errors='coerce'
)
df

,App,Category,Rating,Reviews,Size,Installs,Cash Type,Price,Age Category,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159.0,19M,10000.0,Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967.0,14M,500000.0,Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510.0,8.7M,5000000.0,Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644.0,25M,50000000.0,Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967.0,2.8M,100000.0,Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10836,Sya9a Maroc - FR,FAMILY,4.5,38.0,53M,5000.0,Free,0,Everyone,Education,"July 25, 2017",1.48,4.1 and up
10837,Fr. Mike Schmitz Audio Teachings,FAMILY,5.0,4.0,3.6M,100.0,Free,0,Everyone,Education,"July 6, 2018",1,4.1 and up
10838,Parkinson Exercices FR,MEDICAL,NaN,3.0,9.5M,1000.0,Free,0,Everyone,Medical,"January 20, 2017",1,2.2 and up
10839,The SCP Foundation DB fr nn5n,BOOKS_AND_REFERENCE,4.5,114.0,Varies with device,1000.0,Free,0,Mature 17+,Books & Reference,"January 19, 2015",Varies with device,Varies with device


In [258]:
df['Price'] = pd.to_numeric(
    df['Price']
    .astype(str)
    .str.replace(r'[^0-9.-]','',regex=True)
    .replace('',None),
    errors='coerce'
)
df

,App,Category,Rating,Reviews,Size,Installs,Cash Type,Price,Age Category,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159.0,19M,10000.0,Free,0.0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967.0,14M,500000.0,Free,0.0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510.0,8.7M,5000000.0,Free,0.0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644.0,25M,50000000.0,Free,0.0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967.0,2.8M,100000.0,Free,0.0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10836,Sya9a Maroc - FR,FAMILY,4.5,38.0,53M,5000.0,Free,0.0,Everyone,Education,"July 25, 2017",1.48,4.1 and up
10837,Fr. Mike Schmitz Audio Teachings,FAMILY,5.0,4.0,3.6M,100.0,Free,0.0,Everyone,Education,"July 6, 2018",1,4.1 and up
10838,Parkinson Exercices FR,MEDICAL,NaN,3.0,9.5M,1000.0,Free,0.0,Everyone,Medical,"January 20, 2017",1,2.2 and up
10839,The SCP Foundation DB fr nn5n,BOOKS_AND_REFERENCE,4.5,114.0,Varies with device,1000.0,Free,0.0,Mature 17+,Books & Reference,"January 19, 2015",Varies with device,Varies with device


In [259]:
#Parse last_updated as datetime
df["Last Updated"] = pd.to_datetime(df["Last Updated"],
                                    errors="coerce")
#dt["hour"] = df["lasr_updated"].df.hour
#dt["month"] = df["last_updated"].dt.month
df

,App,Category,Rating,Reviews,Size,Installs,Cash Type,Price,Age Category,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159.0,19M,10000.0,Free,0.0,Everyone,Art & Design,2018-01-07,1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967.0,14M,500000.0,Free,0.0,Everyone,Art & Design;Pretend Play,2018-01-15,2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510.0,8.7M,5000000.0,Free,0.0,Everyone,Art & Design,2018-08-01,1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644.0,25M,50000000.0,Free,0.0,Teen,Art & Design,2018-06-08,Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967.0,2.8M,100000.0,Free,0.0,Everyone,Art & Design;Creativity,2018-06-20,1.1,4.4 and up
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10836,Sya9a Maroc - FR,FAMILY,4.5,38.0,53M,5000.0,Free,0.0,Everyone,Education,2017-07-25,1.48,4.1 and up
10837,Fr. Mike Schmitz Audio Teachings,FAMILY,5.0,4.0,3.6M,100.0,Free,0.0,Everyone,Education,2018-07-06,1,4.1 and up
10838,Parkinson Exercices FR,MEDICAL,NaN,3.0,9.5M,1000.0,Free,0.0,Everyone,Medical,2017-01-20,1,2.2 and up
10839,The SCP Foundation DB fr nn5n,BOOKS_AND_REFERENCE,4.5,114.0,Varies with device,1000.0,Free,0.0,Mature 17+,Books & Reference,2015-01-19,Varies with device,Varies with device


In [260]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   App           10841 non-null  object        
 1   Category      10841 non-null  object        
 2   Rating        9367 non-null   float64       
 3   Reviews       10841 non-null  float64       
 4   Size          10841 non-null  object        
 5   Installs      10840 non-null  float64       
 6   Cash Type     10840 non-null  object        
 7   Price         10840 non-null  float64       
 8   Age Category  10840 non-null  object        
 9   Genres        10841 non-null  object        
 10  Last Updated  10840 non-null  datetime64[ns]
 11  Current Ver   10833 non-null  object        
 12  Android Ver   10838 non-null  object        
dtypes: datetime64[ns](1), float64(4), object(8)
memory usage: 1.1+ MB


## Funtion to convert size to MB

In [261]:
def convert_size(size):
    if size == "Varies with device":
        return None
    elif "M" in size:
        return float(size.replace("M", ""))
    elif "K" in size:
        return float(size.replace("K", "")) / 1024
    else:
        return None


## Apply convertion

In [262]:
df["Size_MB"] = df["Size"].apply(convert_size)

* Optional: drop original column
* df = df.drop(columns=["size])

In [263]:
print(df.head())

                                                 App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

    Reviews  Size    Installs Cash Type  Price Age Category  \
0     159.0   19M     10000.0      Free    0.0     Everyone   
1     967.0   14M    500000.0      Free    0.0     Everyone   
2   87510.0  8.7M   5000000.0      Free    0.0     Everyone   
3  215644.0   25M  50000000.0      Free    0.0         Teen   
4     967.0  2.8M    100000.0      Free    0.0     Everyone   

                      Genres Last Updated         Current Ver   Android Ver  \
0               Art & Design   2018-01-07               1.0.0

In [264]:
df.drop(columns=['Size'],inplace=True)

In [265]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   App           10841 non-null  object        
 1   Category      10841 non-null  object        
 2   Rating        9367 non-null   float64       
 3   Reviews       10841 non-null  float64       
 4   Installs      10840 non-null  float64       
 5   Cash Type     10840 non-null  object        
 6   Price         10840 non-null  float64       
 7   Age Category  10840 non-null  object        
 8   Genres        10841 non-null  object        
 9   Last Updated  10840 non-null  datetime64[ns]
 10  Current Ver   10833 non-null  object        
 11  Android Ver   10838 non-null  object        
 12  Size_MB       8829 non-null   float64       
dtypes: datetime64[ns](1), float64(5), object(7)
memory usage: 1.1+ MB


In [266]:
df.isnull()

,App,Category,Rating,Reviews,Installs,Cash Type,Price,Age Category,Genres,Last Updated,Current Ver,Android Ver,Size_MB
0,False,False,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10836,False,False,False,False,False,False,False,False,False,False,False,False,False
10837,False,False,False,False,False,False,False,False,False,False,False,False,False
10838,False,False,True,False,False,False,False,False,False,False,False,False,False
10839,False,False,False,False,False,False,False,False,False,False,False,False,True


In [267]:
df.isnull().sum()

App                0
Category           0
Rating          1474
Reviews            0
Installs           1
Cash Type          1
Price              1
Age Category       1
Genres             0
Last Updated       1
Current Ver        8
Android Ver        3
Size_MB         2012
dtype: int64

* Rating - 1474(filling)
* Installs - 1(drop)
* cash Type - 1(drop)
* Price - 1(drop)
* Age Category - 1(drop)
* Last Updated -1 (drop)
* current var - 8(drop)
* Android var -3(drp)
* Size_MB - 2012(filling)


In [269]:
df['Rating'].fillna(df['Rating'].median(),inplace=True)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1112\3358796763.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Rating'].fillna(df['Rating'].median(),inplace=True)


,App,Category,Rating,Reviews,Installs,Cash Type,Price,Age Category,Genres,Last Updated,Current Ver,Android Ver,Size_MB
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159.0,10000.0,Free,0.0,Everyone,Art & Design,2018-01-07,1.0.0,4.0.3 and up,19.0
1,Coloring book moana,ART_AND_DESIGN,3.9,967.0,500000.0,Free,0.0,Everyone,Art & Design;Pretend Play,2018-01-15,2.0.0,4.0.3 and up,14.0
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510.0,5000000.0,Free,0.0,Everyone,Art & Design,2018-08-01,1.2.4,4.0.3 and up,8.7
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644.0,50000000.0,Free,0.0,Teen,Art & Design,2018-06-08,Varies with device,4.2 and up,25.0
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967.0,100000.0,Free,0.0,Everyone,Art & Design;Creativity,2018-06-20,1.1,4.4 and up,2.8
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10836,Sya9a Maroc - FR,FAMILY,4.5,38.0,5000.0,Free,0.0,Everyone,Education,2017-07-25,1.48,4.1 and up,53.0
10837,Fr. Mike Schmitz Audio Teachings,FAMILY,5.0,4.0,100.0,Free,0.0,Everyone,Education,2018-07-06,1,4.1 and up,3.6
10838,Parkinson Exercices FR,MEDICAL,4.3,3.0,1000.0,Free,0.0,Everyone,Medical,2017-01-20,1,2.2 and up,9.5
10839,The SCP Foundation DB fr nn5n,BOOKS_AND_REFERENCE,4.5,114.0,1000.0,Free,0.0,Mature 17+,Books & Reference,2015-01-19,Varies with device,Varies with device,NaN


In [271]:
df['Size_MB'].fillna(df['Size_MB'].median(),inplace=True)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1112\2110190827.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Size_MB'].fillna(df['Size_MB'].median(),inplace=True)


,App,Category,Rating,Reviews,Installs,Cash Type,Price,Age Category,Genres,Last Updated,Current Ver,Android Ver,Size_MB
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159.0,10000.0,Free,0.0,Everyone,Art & Design,2018-01-07,1.0.0,4.0.3 and up,19.0
1,Coloring book moana,ART_AND_DESIGN,3.9,967.0,500000.0,Free,0.0,Everyone,Art & Design;Pretend Play,2018-01-15,2.0.0,4.0.3 and up,14.0
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510.0,5000000.0,Free,0.0,Everyone,Art & Design,2018-08-01,1.2.4,4.0.3 and up,8.7
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644.0,50000000.0,Free,0.0,Teen,Art & Design,2018-06-08,Varies with device,4.2 and up,25.0
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967.0,100000.0,Free,0.0,Everyone,Art & Design;Creativity,2018-06-20,1.1,4.4 and up,2.8
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10836,Sya9a Maroc - FR,FAMILY,4.5,38.0,5000.0,Free,0.0,Everyone,Education,2017-07-25,1.48,4.1 and up,53.0
10837,Fr. Mike Schmitz Audio Teachings,FAMILY,5.0,4.0,100.0,Free,0.0,Everyone,Education,2018-07-06,1,4.1 and up,3.6
10838,Parkinson Exercices FR,MEDICAL,4.3,3.0,1000.0,Free,0.0,Everyone,Medical,2017-01-20,1,2.2 and up,9.5
10839,The SCP Foundation DB fr nn5n,BOOKS_AND_REFERENCE,4.5,114.0,1000.0,Free,0.0,Mature 17+,Books & Reference,2015-01-19,Varies with device,Varies with device,14.0


In [272]:
df.isnull().sum()

App             0
Category        0
Rating          0
Reviews         0
Installs        1
Cash Type       1
Price           1
Age Category    1
Genres          0
Last Updated    1
Current Ver     8
Android Ver     3
Size_MB         0
dtype: int64

In [ ]:
df.dropna(subset=['Installs','Cash Type','Price','Age Category','Last Updated','Current Ver','Android Ver'],inplace=True)

In [276]:
df.isnull().sum()

App             0
Category        0
Rating          0
Reviews         0
Installs        0
Cash Type       0
Price           0
Age Category    0
Genres          0
Last Updated    0
Current Ver     0
Android Ver     0
Size_MB         0
dtype: int64

In [278]:
df.shape

(10829, 13)

## Identifying Duplicates

In [280]:
df.duplicated().sum()

np.int64(483)